server/src/main/java/org/opensearch/index/mapper/DerivedFieldType.java

```diff
@@ -0,0 +1,363 @@
+        return new DerivedFieldQuery(
+            query,
+            valueFetcher,
+            context.lookup(),
+            indexableFieldGenerator,
+            typeFieldMapper.mappedFieldType.indexAnalyzer()
+        );
+    }
+
+    @Override
+    public Query multiPhraseQuery(TokenStream stream, int slop, boolean enablePositionIncrements, QueryShardContext context)
+        throws IOException {
+        Query query = typeFieldMapper.mappedFieldType.multiPhraseQuery(stream, slop, enablePositionIncrements, context);
+        DerivedFieldValueFetcher valueFetcher = new DerivedFieldValueFetcher(getDerivedFieldLeafFactory(context));
+        return new DerivedFieldQuery(
+            query,
+            valueFetcher,
+            context.lookup(),
+            indexableFieldGenerator,
+            typeFieldMapper.mappedFieldType.indexAnalyzer()
+        );
+    }
+
+    @Override
+    public Query phrasePrefixQuery(TokenStream stream, int slop, int maxExpansions, QueryShardContext context) throws IOException {
+        Query query = typeFieldMapper.mappedFieldType.phrasePrefixQuery(stream, slop, maxExpansions, context);
+        DerivedFieldValueFetcher valueFetcher = new DerivedFieldValueFetcher(getDerivedFieldLeafFactory(context));
+        return new DerivedFieldQuery(
+            query,
+            valueFetcher,
+            context.lookup(),
+            indexableFieldGenerator,
+            typeFieldMapper.mappedFieldType.indexAnalyzer()
+        );
+    }
+
+    @Override
+    public SpanQuery spanPrefixQuery(String value, SpanMultiTermQueryWrapper.SpanRewriteMethod method, QueryShardContext context) {
+        throw new IllegalArgumentException(
+            "Can only use span prefix queries on text fields - not on [" + name() + "] which is of type [" + typeName() + "]"
+        );
+    }
+
+    @Override
+    public Query distanceFeatureQuery(Object origin, String pivot, float boost, QueryShardContext context) {
+        Query query = typeFieldMapper.mappedFieldType.distanceFeatureQuery(origin, pivot, boost, context);
+        DerivedFieldValueFetcher valueFetcher = new DerivedFieldValueFetcher(getDerivedFieldLeafFactory(context));
+        return new DerivedFieldQuery(
+            query,
+            valueFetcher,
+            context.lookup(),
+            indexableFieldGenerator,
+            typeFieldMapper.mappedFieldType.indexAnalyzer()
+        );
+    }
+
+    @Override
+    public Query existsQuery(QueryShardContext context) {
+        throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] does not support exist queries");
+    }
+
+    @Override
+    public boolean isAggregatable() {
+        return false;
+    }
+
+    private DerivedFieldScript.LeafFactory getDerivedFieldLeafFactory(QueryShardContext context) {
+        if (!context.documentMapper("").sourceMapper().enabled()) {
```

Comment:
Should I remove this check? the scripts will work for the cases when source is disabled and script is just making use of docvalues and not source.

Reply:
I think it's safe to leave it in for now.

Removing the check would add functionality. Let's make progress on the current implementation and can expand as needed.